src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs

```diff
@@ -423,145 +408,154 @@ public void Abort()
-                            }).Unwrap();
-                            await _currentOperation;
-                        }
-
-                        if (read == 0)
-                        {
-                            break;
-                        }
-
-                        Input.Writer.Advance(read);
+                        doneReading = await ReadToInputPipe();
                     }
-                    finally
+
+                    // Write from the output pipe
+                    if (await WriteFromOutputPipe())
                     {
-                        Input.Writer.Commit();
-                        _inputHandle.Dispose();
+                        break;
                     }
 
-                    var result = await Input.Writer.FlushAsync();
-
-                    if (result.IsCompleted || result.IsCancelled)
+                    if (_upgradeTcs != null)
                     {
-                        break;
+                        upgrade = true;
                     }
                 }
 
-                Input.Writer.Complete();
+                // If we upgraded the stream, we can switch to simultaneously reading and writing
+                // create two new tasks and await them finishing.
+                // We don't want to Complete the Reader and Writer as they will be reused in
+                // the other tasks.
+                if (upgrade && _upgradeTcs.TrySetResult(null))
+                {
+                    _readWebsocketTask = ReadWebSockets();
+                    _writeWebsocketTask = WriteWebSockets();
+                }
+                else
+                {
+                    Output.Reader.Complete();
+                    Input.Writer.Complete();
+                }
+
             }
             catch (Exception ex)
             {
                 Input.Writer.Complete(ex);
+                Output.Reader.Complete(ex);
             }
         }
 
-        public void StartWritingResponseBody()
+        // returns true if we are done reading.
+        private async Task<bool> ReadToInputPipe()
         {
-            if (_writingTask == null)
+            var doneReading = false;
+            var wb = Input.Writer.GetMemory(MinAllocBufferSize);
+            _inputHandle = wb.Retain(true);
+
+            try
+            {
+                var read = await UnsafeReadAsync(wb.Length);
+
+                if (read == 0)
+                {
+                    doneReading = true;
+                }
+
+                Input.Writer.Advance(read);
+            }
+            catch (COMException comException)
+            {
+                // If the exception was an aborted read operation, ignore it
+                // as we were the one to invoke it.
+                // E_OPERATIONABORTED == 0x800703e3 == -2147023901
+                // TODO check if this error code can be reported through OnAsyncCompletion other than by calling CancelIo.
+                if (comException.ErrorCode != -2147023901)
```

Comment:
Could we check the error code before calling `_operation.Complete(0x800703e3, 0)`? I don't like potentially throwing an exception for every write.

Reply:
We can check inside of Complete(). Currently GetResult is the one to throw the exception.